lib/topics_bulk_action.rb

```diff
@@ -81,6 +81,24 @@ def dismiss_posts
     @changed_ids.concat @topic_ids
   end
 
+  def dismiss_topics
+    rows = Topic.where(id: @topic_ids)
+      .joins("LEFT JOIN topic_users ON topic_users.topic_id = topics.id AND topic_users.user_id = #{@user.id}")
+      .where("topics.created_at >= ?", dismiss_topics_since_date)
+      .where("topic_users.last_read_post_number IS NULL")
+      .where("topics.archetype <> ?", Archetype.private_message)
+      .order("topics.created_at DESC")
+      .limit(SiteSetting.max_new_topics).map do |topic|
+      {
+        topic_id: topic.id,
+        user_id: @user.id,
+        created_at: Time.zone.now
+      }
+    end
+    DismissedTopicUser.insert_all(rows) if rows.present?
```

Comment:
do we want to allow for duplicates and race conditions here?

Reply:
@SamSaffron based on the docs https://apidock.com/rails/v6.0.0/ActiveRecord/Persistence/ClassMethods/insert_all and other usages of insert_all in our code, it skips duplicates by default, so I don't think there is anything we need to do here.

https://github.com/discourse/discourse/blob/a1ea4a3b4ef0862c69207f4a08661ffea66c34c8/app/models/topic_thumbnail.rb#L24-L30